File: tutorials/b1-checkpoint-restart/cxx/run-cmaes.cpp
// In this example, we demonstrate how a Korali experiment can
// be resumed from any point (generation). This is a useful feature
// for continuing jobs after an error, or to fragment big jobs into
// smaller ones that can better fit a supercomputer queue.

#include "korali.hpp"
#include "model/model.h"

int main(int argc, char* argv[])
{
 // First, we run a simple Korali experiment.
 auto k = Korali::Engine();

 k["Problem"]["Type"] = "Optimization";
 k["Problem"]["Objective"] = "Maximize";

 k["Variables"][0]["Name"] = "X";
 k["Variables"][0]["Lower Bound"] = -10.0;
 k["Variables"][0]["Upper Bound"] = +10.0;

 k["Solver"]["Type"] = "CMAES";
 k["Solver"]["Sample Count"] = 16;

 k["General"]["Max Generations"] = 500;
 k["General"]["Console Output"]["Frequency"] = 10;
 k["General"]["Results Output"]["Frequency"] = 10;

 k.setModel([](Korali::Model& d) { model(d.getVariables(), d.getResults()); });
 k.run();

 printf("\n\nRestarting now:\n\n");

 // Now we loadState() to resume the same experiment from generation 10
 k.loadState("_korali_result/s00010.json");

 k.run();
}

File: features/partial.runs/run-cmaes.py
#!/usr/bin/env python3

# In this example, we demonstrate how a Korali experiment can
# be resumed from any point (generation). This is a useful feature
# for continuing jobs after an error, or to fragment big jobs into
# smaller ones that can better fit a supercomputer queue.

# First, we run a simple Korali experiment.

import sys
sys.path.append('_model')
from model import *
import korali

k = korali.Engine()
e = korali.Experiment()

e["Problem"]["Type"] = "Optimization"
e["Problem"]["Objective Function"] = model

e["Solver"]["Type"] = "Optimizer/CMAES"
e["Solver"]["Population Size"] = 5
e["Solver"]["Termination Criteria"]["Max Generations"] = 50

e["Variables"][0]["Name"] = "X"
e["Variables"][0]["Lower Bound"] = -10.0
e["Variables"][0]["Upper Bound"] = +10.0

e["Console Output"]["Frequency"] = 25

print('------------------------------------------------------')
print('Now running first 50 generations...')
print('------------------------------------------------------')

k.run(e)

print('------------------------------------------------------')
print('Now running last 50 generations...')
print('------------------------------------------------------')

e["Solver"]["Termination Criteria"]["Max Generations"] = 100
k.resume(e)
